feat: functional ground services for efb

Saschl opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->
Fixes #[issue_no]

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
Hooked up the UI with the corresponding SimVars to make all the basic ground services working.

I have not touched any design stuff for now, please tell me if needed.

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->

## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->
Requires #3270 to be merged first as the the SimVar framework is being used

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub):

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
Open the ground page and check that every item triggers the correct ground service.

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

sidnov commented:
Already? :O

Benjozork commented on A32NX/SimObjects/AirPlanes/Asobo_A320_NEO/model/A320_NEO_INTERIOR.xml:
```diff
@@ -43,7 +43,7 @@
 
             <UseTemplate Name="ASOBO_GT_Component_Emissive_Gauge">
-                <EMISSIVE_CODE>0.8</EMISSIVE_CODE>
+                <EMISSIVE_CODE>1</EMISSIVE_CODE>
```
This is too bright for contrast in my testing.

Benjozork replied:
Nevermind - this will disappear once my PR is merged and this is rebased

pepperoni505 reviewed (commented):
What modifications did you make to the glTF file?

Saschl commented:
Sorry for the confusion :D

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,105 @@
+
+    toggleGroundAction(action: GroundServices) {
+        return (() =>{
+            setSimVar(action,"1","boolean");
```
Please fix all spacing issues throughout the code.

```suggestion
            setSimVar(action, "1",  "boolean");
```

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,105 @@
+
+    togglePushback() {
+            const tugActive = this.state.tugStatus;
+            //setSimVar(GroundServices.PUSHBACK_TURN,this.getTugHeading(0)* 11930465,"number");
```
Remove commented code.

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,105 @@
+            this.setState ({
+                tugStatus: !tugActive
+            });
+
```
```suggestion
```

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,105 @@
+    }
+
     render() {
+
```
```suggestion
```

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,105 @@
+    TOGGLE_FUEL="K:REQUEST_FUEL_KEY"
+}
+
+
```
```suggestion
```

boierito commented:
A nice addition would be that when a button is pressed, such as the catering one, it goes from green to red for a second and returns to its normal state. An idea to know if the button was actually pressed. Cheers and good job!

stescouse commented:
Category: QA Tester
Name: Discord User stescouse#7271
Date of testing: 01.2.2021
Version of the sim: 1.12.13.0
PR Tested: #3264
Tier of testing: 1
Changes to observe: functional ground services for EFB

Testing technique: TryOut
Steps you've made to test the Pull Request:

1. Spawned C&D at EDDB.
2. Loaded flight plan in from Simbrief.
3. Selected Ground page on EFB.
4. Selected Jetway, Baggage, Catering, and Fuel to summon services.
5. Selected all above services to dismiss.


Flight notes (If required: None

Issues (If Any): None


Overall Rating: Good
Conclusions of the testing: All services acted as expected.  One suggestion would be to add Power ground service just for completion even though present in most gates already.

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,100 @@
+    }
+
+    setTugHeading(action: GroundServices, value: number, type: string) {
+        return (() =>{
```
```suggestion
        return (() => {
```

Still some spacing issues

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,100 @@
+            const tugActive = this.state.tugStatus;
+            setSimVar(GroundServices.TOGGLE_PUSHBACK, 1, "boolean");
+            setSimVar(GroundServices.PUSHBACK_TURN, this.getTugHeading(0)* 11930465, "number");
+            this.setState ({
```
```suggestion
            this.setState({
```

Benjozork commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,100 @@
+    TOGGLE_CARGO="K:REQUEST_LUGGAGE",
+    TOGGLE_CATERING="K:REQUEST_CATERING",
+    TOGGLE_FUEL="K:REQUEST_FUEL_KEY"
+}
```
```suggestion
}

```

Benjozork reviewed (changes requested):
I don't understand why the lint is not picking this up

Saschl commented:
Bruh. Sorry, I will fix them

Benjozork commented:
No worries, the lint is broken. Fixing it.

Saschl commented:
I'll make this a draft as I discovered some issues with the pushback. Should be fixable though

mattsains commented:
wow, I have been struggling to get the most basic functionality on this page for a few hours, I can't believe it's done already! Nice work

mattsains commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,142 @@
+        return (currentHeading  + value) % 360;
+    }
+
+    togglePushback() {
```
I think there's an issue here where if you hit the stop button twice, this.state.tugActive will be set to true, but the tug won't be moving the aircraft. I recommend accepting a boolean parameter for the desired tug state, and not toggling the state parameter

Saschl replied:
Good point, thanks! Will adapt this :)

mattsains commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,142 @@
+                    <div id="stop"
+                         onMouseDown={(e) => this.handleClick(() => this.state.tugActive ? this.togglePushback() : {}, e)}
+                         onMouseUp={() => this.unselectButton()}
+                         className={this.state.activeButton === "stop" ? "stop selected" : "stop"}><IconHandStop/>
```
I saw in the discord that you were worried about how wordy the code for the selected style was. What about a helper function like this:

```
applySelected(className: string) {
   return className + (this.state.activeButton === className ? ' selected' : '');
}
```

and then here you could have:

```
className={this.applySelected('stop')}
```

Saschl replied:
Heh that makes sense.

mattsains commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,142 @@
-                    <div className="down-right"><IconCornerDownRight/></div>
+                    <div id="stop"
+                         onMouseDown={(e) => this.handleClick(() => this.state.tugActive ? this.togglePushback() : {}, e)}
+                         onMouseUp={() => this.unselectButton()}
```
I was surprised to see the onMouseUp behaviour here. I think you're using the selected class as a pressed animation.

When I wrote the selected class I was thinking you would toggle the button's selected state, so that while you're pushing back, the back button would be highlighted so you would know what the efb is commanding the tug to do.

Saschl replied:
Exactly. Your way makes more sense though, I'll try to adapt it this way :) If you have an idea for classes for the pressed "animation" I'd be happy to add, but I will look as well

mattsains replied:
yeah I think a pressed animation would be good as well because it's pretty non-obvious that you've pressed a button on the efb. Whatever you think is good is ok with me

Saschl commented:
> wow, I have been struggling to get the most basic functionality on this page for a few hours, I can't believe it's done already! Nice work

Thanks for all the feedback!

mattsains commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,171 @@
+        }
+    }
+
+    handleClick(callBack: () => void, event: React.MouseEvent) {
+        if (!this.state.tugActive) {
+            if (event.currentTarget.id === this.state.activeButton) {
+                this.setState({
+                    activeButton: "none"
+                });
+            } else if (!event.currentTarget.id.includes("down-")) {
+                this.setState({
+                    activeButton: event.currentTarget.id
+                });
+            }
+            callBack();
+        } else {
+            if (event.currentTarget.id.match("(down)|(stop)")) {
+                this.setState({
+                    activeButton: event.currentTarget.id
+                });
+                callBack();
+            }
+        }
+    }
```
one issue with this code (and probably the `state.activeButton` field, is that it's possible to call more than one ground service at the same time (eg., both baggage and catering). If you do that, only the last button clicked will show as selected.

mattsains commented on src/instruments/src/EFB/Ground/Ground.tsx:
```diff
@@ -17,47 +17,171 @@
+                this.setState({
+                    activeButton: "none"
+                });
+            } else if (!event.currentTarget.id.includes("down-")) {
+                this.setState({
+                    activeButton: event.currentTarget.id
+                });
+            }
```
the result of this condition is that if you click down-left or down-right, the pushback will start, but the button won't show as active.

Also, after testing this in the sim, I noticed that if the first button you press is down-left or right, the tug will actually push you straight back until you press it again.

I think that if you press down-left or right before back, we should start the tug pushing you in the desired direction rather than requiring the user to select down first.

Saschl commented:
Hey @mattsains sorry for the late reply. I read your comments and was (hopefully) able to address all of this :) The pushback directions are a bit tricky tough, as I changed the function to command the direction also when not being active. But the tug pushes straight  anyway, I need to investigate a bit more. I also added a sync with the actual ground services in game (not pushed yet). Will let you know once I’m confident it works 💯
